### PR TITLE
Add restaurant geo location to the system

### PR DESCRIPTION
**Pull request by junydania**
## What was done?
      - Implemented geocoder gem to convert the restaurant's full address into geo-coordinates

## What we learnt?
     - How to use the geo coder gem
     - How to use cucumber to check for floats

# PT link
     - https://www.pivotaltracker.com/story/show/145100177

### CODE REVIEW

**Comment by tochman** [comment]
There are conflicts on this one that need to be fixed. @andycarnahan can you help them with that?

**Comment by junydania** [comment]
Conflict has been resolved.

File: spec/models/restaurant_spec.rb [hunk 1]
```diff
@@ -7,6 +7,11 @@
     it { is_expected.to have_db_column :address }
     it { is_expected.to have_db_column :email }
     it { is_expected.to have_db_column :phone }
+    it { is_expected.to have_db_column :city  }
+    it { is_expected.to have_db_column :state  }
+    it { is_expected.to have_db_column :country  }
+    it { is_expected.to have_db_column :latitude }
+    it { is_expected.to have_db_column :longitude }
     it { is_expected.to have_db_column :description }
   end
```
Comment (tochman) [inline spec/models/restaurant_spec.rb:18]: IMO you need unit specs for `full_address` and `address_city_state_country_present?` and `address_city_state_country_changed?` (unless you make the last ones `private`)

File: app/models/restaurant.rb [hunk 2]
```diff
@@ -1,4 +1,23 @@
...
+  def address_city_state_country_changed?
+    address_changed? || city_changed? || state_changed? || country_changed?
+  end
+
+  def address_city_state_country_present?
+    address.present? || city.present? || state.present? || country.present?
+  end
+
+  geocoded_by :full_address, latitude: :latitude, longitude: :longitude
+  after_validation  :geocode, if:  :address_city_state_country_present? || :address_city_state_country_changed?
```
Comment (tochman) [inline app/models/restaurant.rb]: I would place this before any methods but after the validations.

File: spec/models/restaurant_spec.rb [hunk 3]
```diff
@@ -25,4 +25,27 @@
       expect(FactoryGirl.create(:restaurant)).to be_valid
     end
   end
+
+  describe 'Geocoder' do
+    before(:all) do
```
Comment (tochman) [inline spec/models/restaurant_spec.rb]: `before` is enough. No need to use `before(:all)` anymore

File: spec/models/restaurant_spec.rb [hunk 4]
```diff
@@ -25,4 +25,27 @@
...
+            'city'      => 'Gothenbourg',
+            'country'   => 'Sweden',
+          }
+        ]
+      )
+    end
+    it "should return latitude" do
+      results = Geocoder.search("Ostrahamngatan 5,")
+      lat = results[0].latitude
+      expect(lat).to eq 57.7093164
```
Comment (tochman) [inline spec/models/restaurant_spec.rb]: This tests Geocoder. You should test the Restaurant object's ability to be geocoded. Does that make sense?

File: spec/models/restaurant_spec.rb [hunk 5]
```diff
@@ -25,4 +25,27 @@
...
+            'latitude' =>   57.7093164,
+            'longitude' =>  11.9663941,
+            'address'   => 'Ostrahamngatan 5,',
+            'state'     => 'Vastra Gotaland',
+            'city'      => 'Gothenbourg',
+            'country'   => 'Sweden',
+          }
+        ]
+      )
+    end
```
Comment (tochman) [inline spec/models/restaurant_spec.rb]: You can probably extract this into `rails_helper`.

**Review by tochman — CHANGES_REQUESTED** [review]

File: spec/models/restaurant_spec.rb (hunk 1 above)
Comment (junydania) [reply spec/models/restaurant_spec.rb:18]: Done

File: app/models/restaurant.rb (hunk 2 above)
Comment (junydania) [reply app/models/restaurant.rb]: Done

File: spec/models/restaurant_spec.rb (hunk 3 above)
Comment (junydania) [reply spec/models/restaurant_spec.rb]: rectified

File: spec/models/restaurant_spec.rb (hunk 5 above)
Comment (junydania) [reply spec/models/restaurant_spec.rb]: would appreciate help, when I moved this block to the rails_helper file, I am unable to make the test pass except this block remains in the spec test file.

**Comment by diraulo** [comment]
@junydania what's the status on this?

**Comment by junydania** [comment]
@diraulo , @tochman complained about this pull request having the same commit with one other pull request of displaying the user on the map.

This pull request was created off develop.

@tochman kindly advise on the next steps for this pull request. Can this be merged while we resolve the other pull request that has the same commit as this?

**Review by tochman — APPROVED** [review]

**Pull request merged**